In [261]:
# Initialize Otter
import otter
grader = otter.Notebook("Assignment 3 - Exploratory Data Analysis And Visualization.ipynb")

### Importing Libraries and Magic Commands

In CSPB 3022, we will be using common Python libraries to help us process data. By convention, we import all libraries at the very top of the notebook. There are also a set of standard aliases that are used to shorten the library names. Below are some of the libraries that you may encounter throughout the course, along with their respective aliases.

In [262]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import plotly.express as px
%matplotlib inline

<a id='verytop'></a>

# Homework 3: Practicing with Pandas

## Due on Gradescope


### Detailed Submission Instructions Are Provided at the end of this Notebook







## Collaboration Policy

Data science is a collaborative activity.  However a key step in learning and retention is **creating solutions on your own.**  

Please see the **Course Syllabus for the Collaboration Policy**.

On the other hand, the following are some **examples of things which would NOT usually be
considered to be cheating**:
 - Working on a HW problem on your own first and then discussing with a classmate a particular part in the problem solution where you are stuck.  After clarifying any questions you should then continue to write your solution independently.
 - Asking someone (or searching online) how a particular construct in the language works.
 - Asking someone (or searching online) how to formulate a particular construct in the language.
 - Asking someone for help in finding an error in your program.  
 - Asking someone why a particular construct does not work as you expected in a given program.
   

To test whether you are truly doing your own work and retaining what you've learned you should be able to easily reproduce from scratch and explain a HW solution that was your own when asked in office hours by an Instructor or on a quiz/exam.   


If you have difficulty in formulating the general solution to a problem on your own, or
you have difficulty in translating that general solution into a program, it is advisable to see
your instructor.

We are here to help!  Visit HW Hours and/or post questions on Piazza!



If while completing this assignment you reference any websites other than those linked in this assignment or provided on Canvas please list those references here:

**External references**:  *list any websites you referenced

## Grading
Grading is broken down into autograded answers and manually graded answers. 

For autograded answers, the results of your code are compared to provided and/or hidden tests.

For manually graded answers you must show and explain all steps.  Graders will evaluate how well you answered the question and/or fulfilled the requirements of the question.


### Score breakdown



Question | Points | Grading Type
--- | --- | ---
Question 1a | 5 | autograded
Question 1b | 5 | autograded
Question 1c | 5 | autograded
Question 2 | 5 | autograded
Question 3a | 5 | autograded
Question 3b | 3 | manual
Question 4ai | 4 | autograded
Question 4aii | 1 | autograded
Question 4bi | 4 | autograded
Question 4bii | 1 | autograded
Question 4c | 4 | autograded
Question 4d | 6 | manual
Question 4e | 2 | manual
|Total | 50

<hr style="border: 5px solid #003262;" />
<hr style="border: 1px solid #fdb515;" />



### Question 1: Elections

**Review**: Let's start by reading in the election dataset from the pandas lectures.

In [263]:
# run this cell
elections = pd.read_csv("data/elections.csv")
elections.head(5)

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

As we saw, we can groupby a specific column, e.g. "Party". It turns out that using some syntax we didn't cover in lecture, we can print out the subframes that result. This isn't something you'll do for any practical purpose. However, it may help you get an understanding of what groupby is actually doing.

An example is given below for elections since 1980.

In [264]:
# run this cell
for n, g in elections.query("Year >= 1980").groupby("Party"):
    print(f"Name: {n}") # by the way this is an "f string", a relatively new and great feature of Python
    display(g)

Name: Citizens


Year       Candidate     Party  Popular vote Result         %
127  1980  Barry Commoner  Citizens        233052   loss  0.270182

Name: Constitution


Year         Candidate         Party  Popular vote Result         %
160  2004  Michael Peroutka  Constitution        143630   loss  0.117542
164  2008     Chuck Baldwin  Constitution        199750   loss  0.152398
172  2016    Darrell Castle  Constitution        203091   loss  0.149640

Name: Democratic


Year        Candidate       Party  Popular vote Result          %
129  1980     Jimmy Carter  Democratic      35480115   loss  41.132848
134  1984   Walter Mondale  Democratic      37577352   loss  40.729429
137  1988  Michael Dukakis  Democratic      41809074   loss  45.770691
140  1992     Bill Clinton  Democratic      44909806    win  43.118485
144  1996     Bill Clinton  Democratic      47400125    win  49.296938
151  2000          Al Gore  Democratic      50999897   loss  48.491813
158  2004       John Kerry  Democratic      59028444   loss  48.306775
162  2008     Barack Obama  Democratic      69498516    win  53.023510
168  2012     Barack Obama  Democratic      65915795    win  51.258484
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515

Name: Green


Year         Candidate  Party  Popular vote Result         %
149  1996       Ralph Nader  Green        685297   loss  0.712721
155  2000       Ralph Nader  Green       2882955   loss  2.741176
156  2004        David Cobb  Green        119859   loss  0.098088
165  2008  Cynthia McKinney  Green        161797   loss  0.123442
170  2012        Jill Stein  Green        469627   loss  0.365199
177  2016        Jill Stein  Green       1457226   loss  1.073699
181  2020    Howard Hawkins  Green        405035   loss  0.255731

Name: Independent


Year         Candidate        Party  Popular vote Result          %
130  1980  John B. Anderson  Independent       5719850   loss   6.631143
143  1992        Ross Perot  Independent      19743821   loss  18.956298
161  2004       Ralph Nader  Independent        465151   loss   0.380663
167  2008       Ralph Nader  Independent        739034   loss   0.563842
174  2016     Evan McMullin  Independent        732273   loss   0.539546

Name: Libertarian


Year         Candidate        Party  Popular vote Result         %
128  1980          Ed Clark  Libertarian        921128   loss  1.067883
132  1984    David Bergland  Libertarian        228111   loss  0.247245
138  1988          Ron Paul  Libertarian        431750   loss  0.472660
139  1992      Andre Marrou  Libertarian        290087   loss  0.278516
146  1996      Harry Browne  Libertarian        485759   loss  0.505198
153  2000      Harry Browne  Libertarian        384431   loss  0.365525
159  2004  Michael Badnarik  Libertarian        397265   loss  0.325108
163  2008          Bob Barr  Libertarian        523715   loss  0.399565
169  2012      Gary Johnson  Libertarian       1275971   loss  0.992241
175  2016      Gary Johnson  Libertarian       4489235   loss  3.307714
180  2020      Jo Jorgensen  Libertarian       1865724   loss  1.177979

Name: Natural Law


Year     Candidate        Party  Popular vote Result         %
148  1996  John Hagelin  Natural Law        113670   loss  0.118219

Name: New Alliance


Year      Candidate         Party  Popular vote Result         %
136  1988  Lenora Fulani  New Alliance        217221   loss  0.237804

Name: Populist


Year Candidate     Party  Popular vote Result         %
141  1992  Bo Gritz  Populist        106152   loss  0.101918

Name: Reform


Year     Candidate   Party  Popular vote Result         %
150  1996    Ross Perot  Reform       8085294   loss  8.408844
154  2000  Pat Buchanan  Reform        448895   loss  0.426819

Name: Republican


Year          Candidate       Party  Popular vote Result          %
131  1980      Ronald Reagan  Republican      43903230    win  50.897944
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
142  1992  George H. W. Bush  Republican      39104550   loss  37.544784
145  1996           Bob Dole  Republican      39197469   loss  40.766036
152  2000     George W. Bush  Republican      50456002    win  47.974666
157  2004     George W. Bush  Republican      62040610    win  50.771824
166  2008        John McCain  Republican      59948323   loss  45.737243
171  2012        Mitt Romney  Republican      60933504   loss  47.384076
173  2016       Donald Trump  Republican      62984828    win  46.407862
179  2020       Donald Trump  Republican      74216154   loss  46.858542

Name: Taxpayers


Year        Candidate      Party  Popular vote Result         %
147  1996  Howard Phillips  Taxpayers        184656   loss  0.192045

Recall that once we've formed groups, we can aggregate each sub-dataframe (a.k.a. group) into a single row using an aggregation function. For example, if we use `.agg(np.mean)` on the groups above, we get back a single DataFrame where each group has been replaced by a single row. In each column for that aggregate row, the value that appears is the average of all values in that group.


In [265]:
elections_after_1980 = elections[elections["Year"] >= 1980][["Party","Popular vote", "%"]]

elections_after_1980.groupby("Party").agg(np.mean)

/var/folders/tz/1w69jxn933j7tdk71hswkyxm0000gn/T/ipykernel_1098/3675187317.py:3: FutureWarning: The provided callable <function mean at 0x1144cd9e0> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  elections_after_1980.groupby("Party").agg(np.mean)


Popular vote          %
Party                                
Citizens      2.330520e+05   0.270182
Constitution  1.821570e+05   0.139860
Democratic    5.452196e+07  47.360184
Green         8.831137e+05   0.767151
Independent   5.480026e+06   5.414298
Libertarian   1.026652e+06   0.830876
Natural Law   1.136700e+05   0.118219
New Alliance  2.172210e+05   0.237804
Populist      1.061520e+05   0.101918
Reform        4.267094e+06   4.417831
Republican    5.419334e+07  47.898650
Taxpayers     1.846560e+05   0.192045

Equivalently we can use one of the shorthand aggregation functions, e.g. `.mean()`: 

In [266]:
elections_after_1980.groupby("Party").mean()

Popular vote          %
Party                                
Citizens      2.330520e+05   0.270182
Constitution  1.821570e+05   0.139860
Democratic    5.452196e+07  47.360184
Green         8.831137e+05   0.767151
Independent   5.480026e+06   5.414298
Libertarian   1.026652e+06   0.830876
Natural Law   1.136700e+05   0.118219
New Alliance  2.172210e+05   0.237804
Populist      1.061520e+05   0.101918
Reform        4.267094e+06   4.417831
Republican    5.419334e+07  47.898650
Taxpayers     1.846560e+05   0.192045

Note that the index of the dataframe returned by an `groupby.agg` call is no longer a set of numeric indices from 0 to N-1. Instead, we see that the index for the example above is now the `Party`. If we want to restore our DataFrame so that `Party` is a column rather than the index, we can use `reset_index`.

In [267]:
elections_after_1980.groupby("Party").mean().reset_index()

Party  Popular vote          %
0       Citizens  2.330520e+05   0.270182
1   Constitution  1.821570e+05   0.139860
2     Democratic  5.452196e+07  47.360184
3          Green  8.831137e+05   0.767151
4    Independent  5.480026e+06   5.414298
5    Libertarian  1.026652e+06   0.830876
6    Natural Law  1.136700e+05   0.118219
7   New Alliance  2.172210e+05   0.237804
8       Populist  1.061520e+05   0.101918
9         Reform  4.267094e+06   4.417831
10    Republican  5.419334e+07  47.898650
11     Taxpayers  1.846560e+05   0.192045

**IMPORTANT NOTE:** Notice that the code above consists of a series of chained method calls. This sort of code is very very common in Pandas programming and in data science in general. Such chained method calls can sometimes go many layers deep, in which case you might consider adding newlines between lines of code for clarity. For example, we could instead write the code above as:

In [268]:
# pandas method chaining
(
elections.query("Year >= 1980")[["Party","Popular vote", "%"]].groupby("Party") 
                               .mean()            ## computes the mean values by party
                               .reset_index()     ## reset to a numerical index
)

Party  Popular vote          %
0       Citizens  2.330520e+05   0.270182
1   Constitution  1.821570e+05   0.139860
2     Democratic  5.452196e+07  47.360184
3          Green  8.831137e+05   0.767151
4    Independent  5.480026e+06   5.414298
5    Libertarian  1.026652e+06   0.830876
6    Natural Law  1.136700e+05   0.118219
7   New Alliance  2.172210e+05   0.237804
8       Populist  1.061520e+05   0.101918
9         Reform  4.267094e+06   4.417831
10    Republican  5.419334e+07  47.898650
11     Taxpayers  1.846560e+05   0.192045

Note that we have surrounded the entire call by a big set of parentheses so that Python doesn't complain about the indentation. An alternative is to use the \ symbol to indicate to Python that your code continues on to the next line!

In [269]:
# pandas method chaining (alternative)
elections.query("Year >= 1980")[["Party","Popular vote", "%"]].groupby("Party") \
                               .mean() \
                               .reset_index()     

Party  Popular vote          %
0       Citizens  2.330520e+05   0.270182
1   Constitution  1.821570e+05   0.139860
2     Democratic  5.452196e+07  47.360184
3          Green  8.831137e+05   0.767151
4    Independent  5.480026e+06   5.414298
5    Libertarian  1.026652e+06   0.830876
6    Natural Law  1.136700e+05   0.118219
7   New Alliance  2.172210e+05   0.237804
8       Populist  1.061520e+05   0.101918
9         Reform  4.267094e+06   4.417831
10    Republican  5.419334e+07  47.898650
11     Taxpayers  1.846560e+05   0.192045

**IMPORTANT NOTE:** You should NEVER solve problems like the one above using loops or list comprehensions. This is slow and also misses the entire point of this part of CSPB 3022. 

Before we continue, we'll print out the election dataset again for your convenience. 

In [270]:
elections.head(5)

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

#### Question 1a
Using `groupby.agg` or one of the shorthand methods (`groupby.min`, `groupby.first`, etc.), create a DataFrame `best_result_percentage_only` showing the overall best result for every party, sorted in decreasing order. 
Your DataFrame should **include only parties which have earned at least 10% of the vote** in some election. 

For example, the first 3 rows of your DataFrame should be:

|Party | %         |
|------|------|
|**Democratic**  | 61.344703 |
|**Republican**  | 60.907806 |
|**Democratic-Republican**   | 57.210122 |

Note that the index is `Party`. In other words, don't use `reset_index`.


A list of named `groupby.agg` shorthand methods is [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#aggregation) (you'll have to scroll down about one page).


In [271]:
best_result_percentage_only = elections[elections["%"] >= 10][["Party", "%"]].groupby("Party").max().sort_values(by = "%", ascending = False)

# put your code above this line
best_result_percentage_only

%
Party                           
Democratic             61.344703
Republican             60.907806
Democratic-Republican  57.210122
National Union         54.951512
Whig                   53.051213
Liberal Republican     44.071406
National Republican    43.796073
Northern Democratic    29.522311
Progressive            27.457433
American               21.554001
Independent            18.956298
Southern Democratic    18.138998
American Independent   13.571218
Constitutional Union   12.639283
Free Soil              10.138474

In [272]:
grader.check("q1a")

q1a results: All test cases passed!

#### Question 1b  
Repeat Question 1a. However, this time, your result should be a DataFrame called `best_result` that shows the overall best result for each party AND includes columns with the corresponding Year, Candidate, Popular Vote and Result.

This question is trickier than Question 1a. Make sure to check the lecture slides if you're stuck! It's very easy to make a subtle mistake that shows Woodrow Wilson and Howard Taft both winning the 2020 election.

Sort in descending order according to the best result.  

For example, the first 3 rows of your table should be:

|Party | Year | Candidate      | Popular Vote | Result | %         |
|------|------|----------------|--------------|--------|-----------|
|**Democratic**  | 1964 | Lyndon Johnson | 43127041      | win   | 61.344703 |
|**Republican**  | 1972 | Richard Nixon | 47168710      | win   | 60.907806 |
|**Democratic-Republican**  | 1824 | Andrew Jackson | 151271      | loss   | 57.210122 |

Note that the index is `Party`. In other words, don't use `reset_index`.


In [273]:
best_result = elections[elections["%"] >= 10][["Party", "Year", "Candidate", "Popular vote", "Result", "%"]]
best_result = best_result.sort_values(by = "%", ascending = False)
best_result = best_result.drop_duplicates(subset="Party", keep="first")
best_result = best_result.sort_values(by = "%", ascending=False)
best_result = best_result.set_index("Party")

# put your code above this line
best_result

Year               Candidate  Popular vote Result  \
Party                                                                      
Democratic             1964          Lyndon Johnson      43127041    win   
Republican             1972           Richard Nixon      47168710    win   
Democratic-Republican  1824          Andrew Jackson        151271   loss   
National Union         1864         Abraham Lincoln       2211317    win   
Whig                   1840  William Henry Harrison       1275583    win   
Liberal Republican     1872          Horace Greeley       2834761   loss   
National Republican    1828       John Quincy Adams        500897   loss   
Northern Democratic    1860      Stephen A. Douglas       1380202   loss   
Progressive            1912      Theodore Roosevelt       4122721   loss   
American               1856        Millard Fillmore        873053   loss   
Independent            1992              Ross Perot      19743821   loss   
Southern Democratic    1860    John C. Breckinridge        848019   loss   
American Independent   1968          George Wallace       9901118   loss   
Constitutional Union   1860               John Bell        590901   loss   
Free Soil              1848        Martin Van Buren        291501   loss   

                               %  
Party                             
Democratic             61.344703  
Republican             60.907806  
Democratic-Republican  57.210122  
National Union         54.951512  
Whig                   53.051213  
Liberal Republican     44.071406  
National Republican    43.796073  
Northern Democratic    29.522311  
Progressive            27.457433  
American               21.554001  
Independent            18.956298  
Southern Democratic    18.138998  
American Independent   13.571218  
Constitutional Union   12.639283  
Free Soil              10.138474

In [274]:
grader.check("q1b")

q1b results: All test cases passed!

Our DataFrame contains a number of parties which have never had a successful presidential run. For example, the 2020 elections included candiates from the Libertarian and Green parties, neither of which have elected a president.

In [275]:
# just run this cell
elections.tail(5)

Year       Candidate        Party  Popular vote Result          %
177  2016      Jill Stein        Green       1457226   loss   1.073699
178  2020    Joseph Biden   Democratic      81268924    win  51.311515
179  2020    Donald Trump   Republican      74216154   loss  46.858542
180  2020    Jo Jorgensen  Libertarian       1865724   loss   1.177979
181  2020  Howard Hawkins        Green        405035   loss   0.255731

Suppose we were conducting an analysis trying to focus our attention on parties that had elected a president. 

The most natural approach is to use `groupby.filter`. This is an incredibly powerful but subtle tool for filtering data.  

See the Pandas `groupby.filter` [documentation](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.filter.html) for documentation on using `groupby.filter`.



The code below accomplishes the task at hand. It does this by creating a function that returns True if and only if a sub-dataframe (a.k.a. group) contains at least one winner. This function in turn uses the [Pandas function "any"](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html).

In [276]:
# just run this cell
def at_least_one_candidate_in_the_frame_has_won(frame):
    """Returns df with rows only kept for parties that have
    won at least one election
    """
    return (frame["Result"] == 'win').any()

winners_only = (
    elections
        .groupby("Party")
        .filter(at_least_one_candidate_in_the_frame_has_won)
)
winners_only.tail(5)

Year        Candidate       Party  Popular vote Result          %
171  2012      Mitt Romney  Republican      60933504   loss  47.384076
173  2016     Donald Trump  Republican      62984828    win  46.407862
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515
179  2020     Donald Trump  Republican      74216154   loss  46.858542

Alternately we could have used a `lambda` function instead of explicitly defining a named function using `def`. 

In [277]:
# just run this cell (alternative)
winners_only = (
    elections
        .groupby("Party")
        .filter(lambda x : (x["Result"] == "win").any())
)
winners_only.tail(5)

Year        Candidate       Party  Popular vote Result          %
171  2012      Mitt Romney  Republican      60933504   loss  47.384076
173  2016     Donald Trump  Republican      62984828    win  46.407862
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515
179  2020     Donald Trump  Republican      74216154   loss  46.858542



For the next question, you'll do a less restrictive filtering of the elections data.

#### Question 1c

Using `filter`, create a DataFrame `major_party_results_since_1988` that includes all election results starting in 1988, but only show a row if the Party it belongs to has earned at least 1% of the popular vote in ANY election since 1988.

For example, in 1988, you should not include the `New Alliance` candidate, since this party has not earned 1% of the vote since 1988. However, you should include the `Libertarian` candidate from 1988 despite only having 0.47 percent of the vote in 1988, because in 2016 and 2020, the Libertarian candidates Gary Johnson and Jo Jorgensen exceeded 1% of the vote.

For example, the first three rows of the table you generate should look like:

|     |   Year | Candidate         | Party       |   Popular vote | Result   |         % |
|----:|-------:|:------------------|:------------|---------------:|:---------|----------:|
| 135 |   1988 | George H. W. Bush | Republican  |       48886597 | win      | 53.5188   |
| 137 |   1988 | Michael Dukakis   | Democratic  |       41809074 | loss     | 45.7707   |
| 138 |   1988 | Ron Paul          | Libertarian |         431750 | loss     |  0.47266  |

*Hint*: The following questions might help you construct your solution. One of the lines should be identical to the `filter` examples shown above.

1) How can we **only** keep rows in the data that are after 1988?
2) What column should we `groupby` to filter out parties that have earned at least 1% of the popular vote in ANY election since 1988?
3) How can we write an aggregation function that takes a subframe (or sub-DataFrame) and returns whether at least 1% of the vote has been earned in that subframe? This may give you a hint about the second question!


In [278]:
major_party_results_since_1988 = elections.query("Year >= 1988")
major_party_results_since_1988 = major_party_results_since_1988.groupby("Party").filter(lambda x: (x["%"] >= 1).any())


# put your code above this line

major_party_results_since_1988.head()

Year          Candidate        Party  Popular vote Result          %
135  1988  George H. W. Bush   Republican      48886597    win  53.518845
137  1988    Michael Dukakis   Democratic      41809074   loss  45.770691
138  1988           Ron Paul  Libertarian        431750   loss   0.472660
139  1992       Andre Marrou  Libertarian        290087   loss   0.278516
140  1992       Bill Clinton   Democratic      44909806    win  43.118485

In [279]:
grader.check("q1c")

q1c results: All test cases passed!

Pandas provides special purpose functions for working with specific common data types such as strings and dates. For example, the code below provides the length of every Candidate's name from our elections dataset.

In [280]:
elections["Candidate"].str.len()

0      14
1      17
2      14
3      17
4      14
       ..
177    10
178    12
179    12
180    12
181    14
Name: Candidate, Length: 182, dtype: int64

### Question 2

Using `.str.split` create a new DataFrame called `elections_with_first_name` with a new column `First Name` that is equal to the Candidate's first name.

See the Pandas `str` [documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.split.html) for documentation on using `str.split`.

Hint: You will need to use `[0]` somewhere in your code.


In [281]:
modElections = pd.DataFrame(elections)
modElections["First Name"] = elections["Candidate"].str.split().str[0]
elections_with_first_name = modElections

# put your code above this line
elections_with_first_name

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             % First Name  
0    57.210122     Andrew  
1    42.789878       John  
2    56.203927     Andrew  
3    43.796073       John  
4    54.574789     Andrew  
..         ...        ...  
177   1.073699       Jill  
178  51.311515     Joseph  
179  46.858542     Donald  
180   1.177979         Jo  
181   0.255731     Howard  

[182 rows x 7 columns]

In [282]:
grader.check("q2")

q2 results: All test cases passed!

### QUESTION 3: Babynames Dataset

We will now download data from the United States Social Security office containing the number of registered names broken down by **year**, **sex**, and **name**. This is often called the Baby Names Data as social security numbers (SSNs) are typically given at birth.

Reading from [SSN Office description](https://www.ssa.gov/oact/babynames/background.html), bolded for readability: 


> All names are from Social Security card applications for **births that occurred in the United States** after 1879. **Note that many people born before 1937 never applied** for a Social Security card, so their names are not included in our data. For others who did apply, our records may not show the place of birth, and again their names are not included in our data.

> **To safeguard privacy, we exclude** from our tabulated lists of names those that would indicate, or would allow the ability to determine, **names with fewer than 5 occurrences** in any geographic area. If a name has less than 5 occurrences for a year of birth in any state, the sum of the state counts for that year will be less than the national count.

> All data are from a **100% sample** of our records on Social Security card applications as of March 2023.


First let's run the following cells to build the DataFrame `baby_names`.

The cells below download the data from the web and extract the data into a DataFrame. There should be a total of 6215834 records.

The code shown here is outside of the scope of CSCI 3022, but you're encouraged to dig into it if you are interested!


### `fetch_and_cache` Helper

The following function downloads and caches data in the `data/` directory and returns the `Path` to the downloaded file. The cell below the function describes how it works. You are not expected to understand this code, but you may find it useful as a reference as a practitioner of data science after the course. 

In [283]:
import requests
from pathlib import Path

def fetch_and_cache(data_url, file, data_dir="data", force=False):
    """
    Download and cache a url and return the file object.
    
    data_url: the web address to download
    file: the file in which to save the results.
    data_dir: (default="data") the location to save the data
    force: if true the file is always re-downloaded 
    
    return: The pathlib.Path to the file.
    """
    data_dir = Path(data_dir)
    data_dir.mkdir(exist_ok=True)
    file_path = data_dir/Path(file)
    if force and file_path.exists():
        file_path.unlink()
    if force or not file_path.exists():
        print('Downloading...', end=' ')
        resp = requests.get(data_url)
        with file_path.open('wb') as f:
            f.write(resp.content)
        print('Done!')
    else:
        import time 
        created = time.ctime(file_path.stat().st_ctime)
        print("Using cached version downloaded at", created)
    return file_path

In Python, a `Path` object represents the filesystem paths to files (and other resources). The `pathlib` module is effective for writing code that works on different operating systems and filesystems. 

To check if a file exists at a path, use `.exists()`. To create a directory for a path, use `.mkdir()`. To remove a file that might be a [symbolic link](https://en.wikipedia.org/wiki/Symbolic_link), use `.unlink()`. 

This function creates a path to a directory that will contain data files. It ensures that the directory exists (which is required to write files in that directory), then proceeds to download the file based on its URL.

The benefit of this function is that not only can you force when you want a new file to be downloaded using the `force` parameter, but in cases when you don't need the file to be re-downloaded, you can use the cached version and save download time.

Below we use `fetch_and_cache` to download the `namesbystate.zip` zip file, which is a compressed directory of CSV files. 

**This might take a little while! Consider stretching.**

In [284]:
data_url = 'https://www.ssa.gov/oact/babynames/state/namesbystate.zip'
namesbystate_path = fetch_and_cache(data_url, 'namesbystate.zip')

Using cached version downloaded at Mon Jan 22 14:13:19 2024


The following cell builds the final full `baby_names` DataFrame. It first builds one DataFrame per state, because that's how the data are stored in the zip file. Here is documentation for [pd.concat](https://pandas.pydata.org/pandas-docs/version/1.2/reference/api/pandas.concat.html) if you want to know more about its functionality. As before, you are not expected to understand this code. 

In [285]:
import zipfile
zf = zipfile.ZipFile(namesbystate_path, 'r')

column_labels = ['State', 'Sex', 'Year', 'Name', 'Count']

def load_dataframe_from_zip(zf, f):
    with zf.open(f) as fh: 
        return pd.read_csv(fh, header=None, names=column_labels)

states = [
    load_dataframe_from_zip(zf, f)
    for f in sorted(zf.filelist, key=lambda x:x.filename) 
    if f.filename.endswith('.TXT')
]

baby_names = states[0]
for state_df in states[1:]:
    baby_names = pd.concat([baby_names, state_df])
baby_names = baby_names.reset_index().iloc[:, 1:]

In [286]:
baby_names.head()

State Sex  Year      Name  Count
0    AK   F  1910      Mary     14
1    AK   F  1910     Annie     12
2    AK   F  1910      Anna     10
3    AK   F  1910  Margaret      8
4    AK   F  1910     Helen      7

The code below creates a table with the frequency of all names from 2020.

In [287]:
# just run this cell
baby_names_2020 = (
    baby_names.query('Year == 2020')
              .groupby("Name")
              .sum()[["Count"]]
              .reset_index()
)
baby_names_2020

Name  Count
0       Aaden     15
1     Aadhira      6
2     Aadhvik      5
3      Aadhya    186
4        Aadi     14
...       ...    ...
8758   Zymere      6
8759    Zymir     75
8760     Zyon    136
8761     Zyra     33
8762    Zyrah      5

[8763 rows x 2 columns]

#### Question 3a). 
Using the `pd.merge` function described in lecture, combine the `baby_names_2020` table with the `elections_with_first_name` table you created earlier to form `presidential_candidates_and_name_popularity`.


In [288]:
presidential_candidates_and_name_popularity = pd.merge(elections_with_first_name, baby_names_2020, left_on = "First Name", right_on = "Name")
# put your code above this line

presidential_candidates_and_name_popularity

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
149  2016    Hillary Clinton             Democratic      65853514   loss   
150  2020       Joseph Biden             Democratic      81268924    win   
151  2020       Donald Trump             Republican      74216154   loss   
152  2020       Jo Jorgensen            Libertarian       1865724   loss   
153  2020     Howard Hawkins                  Green        405035   loss   

             % First Name     Name  Count  
0    57.210122     Andrew   Andrew   6049  
1    42.789878       John     John   8251  
2    56.203927     Andrew   Andrew   6049  
3    43.796073       John     John   8251  
4    54.574789     Andrew   Andrew   6049  
..         ...        ...      ...    ...  
149  48.521539    Hillary  Hillary     20  
150  51.311515     Joseph   Joseph   8458  
151  46.858542     Donald   Donald    412  
152   1.177979         Jo       Jo      6  
153   0.255731     Howard   Howard    131  

[154 rows x 9 columns]

In [289]:
grader.check("q3a")

q3a results: All test cases passed!

<!-- BEGIN QUESTION -->

#### Question 3b).  

i).  Which historical presidential candidate first name was the most popular in 2020? 

ii).  What 3 historical presidential candidate first names were tied for the least popular in 2020 according to this presidential_candidates_and_name_popularity table? 

Note: Here you'll observe a common problem in data science -- one of the least popular names is actually due to the fact that one recent president was so commonly known by his nickname that he appears named as such in the database from which you pulled election results.

In [290]:
most_popular_firstname = presidential_candidates_and_name_popularity.sort_values(by = "Count", ascending = False)
most_popular_firstname = most_popular_firstname.head(1)
most_popular_firstname = most_popular_firstname["First Name"]

# put your code to calculate the most popular first name above this line, and output the first name
most_popular_firstname



77    William
Name: First Name, dtype: object

In [291]:
least_popular_firstnames = presidential_candidates_and_name_popularity.sort_values(by = "Count", ascending = True)
least_popular_firstnames = least_popular_firstnames.head(1)
least_popular_firstnames = least_popular_firstnames["First Name"]

# put your code to calculate a series with the least popular names above this line
least_popular_firstnames


126    Bill
Name: First Name, dtype: object

<!-- END QUESTION -->

### QUESTION 4: Evidence for Money Ball

In the book [MoneyBall](https://en.wikipedia.org/wiki/Moneyball), Michael Lewis documented the introduction of statistics and data science in selecting players for the Oakland A's. in this problem, we're going to use [Sean Lahman's Baseball Database](http://seanlahman.com/baseball-archive/statistics) which contains the "complete batting and pitching statistics from 1871 to 2013, plus fielding statistics, standings, team stats, managerial records, post-season data, and more. For more details, please [read the documentation](http://seanlahman.com/files/database/readme2012.txt)."

We're going to use two data sets, `teams.csv` and `salaries.csv`. 



Read in the `teams.csv` data into a DataFrame and the `salaries.csv` into another DataFrame and look at the first 5 rows of each. 



In [292]:
teams_df=pd.read_csv("data/teams.csv")

In [293]:
salaries_df=pd.read_csv("data/salaries.csv")

Both files include columns `yearID` and `teamID`. 

The `teamID` field is a categorical field indicating the team names. 

The `yearID` field is a numeric field indicating the year for the data in that row. 

The `teams.csv` file also contains a field `W`, which indicates the total *wins* for that team in that year. 

The `salaries.csv` file contains the salary of individual players. Each player is listed on a distinct line.

#### Question 4ai)- Checking the number of wins in 1998


Create a Dataframe `w1998` with index `teamID` and columns `name` and `wins` (the total number of wins per team in 1998). The team names in the index should be sorted alphabetically from ANA to TOR. 

For example, the first 3 rows of your DataFrame should be:

|teamID | name         |wins |
|------|------|--------|
|**ANA**  | Anaheim Angels |85|
|**ARI**  | Arizona Diamondbacks | 65|
|**ATL**   | Atlanta Braves| 106|

Be sure to rename the "W" column to "wins".

Note that the index is `teamID`. 



In [294]:

...
# put your code above this line



w1998

NameError: name 'w1998' is not defined

In [ ]:
grader.check("q4ai")


#### QUESTION 4aii).

How many games did the Oakland A's win in 1998?

In [ ]:
oakwins= ...
oakwins

In [ ]:
grader.check("q4aii")

Run the code below to produce a bar plot showing each team's total wins in 1998.

In [ ]:
w1998.plot(kind='bar', figsize=(8,4)).set(ylabel='Wins', xlabel = 'Team');

#### Question 4bi - Examining the payroll in 1998



Create a dataframe `p1998` with index `teamID` and column `payroll` (the total salary per team in 1998). 

The team names in the index should be sorted alphabetically from ANA to TOR.

For example, the first 3 rows of your DataFrame should be:

|teamID | payroll         |
|------|------|
|**ANA**  | 41281000|
|**ARI**  |  32347000|
|**ATL**   | 61186000|




In [ ]:
# your code here
...
# end your code



p1998

In [ ]:
grader.check("q4bi")

#### Question 4bii:

What was the Oakland A's payroll in 1998?


In [ ]:
o_payroll = ...

o_payroll

In [ ]:
grader.check("q4bii")

#### QUESTION 4c:   



Merge your dataframes from parts 4a and 4b to create a new dataframe called `df_1998` with the following columns:

`teamID`, `name` `wins` (total number of wins per team in 1998) and `payroll` (total team payroll in 1998)


 


In [ ]:


...
# end your code

df_1998

In [ ]:
grader.check("q4c")

<!-- BEGIN QUESTION -->

#### Question 4di).  What's up with Oakland? 


In this problem, you're going to produce scatter plots to confirm the intuition that the data science approach that Oakland adopted changed their efficiency (wins per dollar spent).


Run the code below to produce a [scatter plot](http://pandas.pydata.org/pandas-docs/stable/visualization.html#scatter-plot) of the payroll (y-axis) *vs* the number of Wins (x-axis) for all teams during the year 1998, using your dataframe `df_1998`. Notice the code below also highlights the datapoint for Oakland in red.  


In [ ]:

df_1998.plot.scatter('wins', 'payroll')

plt.title('Year 1998')
plt.plot(df_1998.loc['OAK','wins'],df_1998.loc['OAK','payroll'], 'ro')
plt.xlim(0,110)
plt.show()

### Question 4dii).

Create two more of these scatterplots (one for 2003 and one for 2013) of wins vs payroll for all teams, and highlight Oakland in red. 


In [ ]:
# your code for 2003 here:
...


In [ ]:
# your code for 2013 here:
...

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

### QUESTION 4e). 

Examining your scatterplots above, what was the effect of introducing statistics and data science in selecting players for the Oakland A's? (i.e. comment on what trend you notice from the graphs regarding the Oakland A's between 1998, 2003 and 2013).      

_Type your answer here, replacing this text._

<!-- END QUESTION -->

<br/><br/>
<hr style="border: 5px solid #003262;" />
<hr style="border: 1px solid #fdb515;" />

## Congratulations! You have finished Homework 3!

If you discussed this assignment with any other students in the class (in a manner that is acceptable as described by the Collaboration policy above) please **include their names** here:

**Collaborators**: *list collaborators here*

### Submission Instructions

Before proceeding any further, **save this notebook.**

After running the `grader.export()` cell provided below, **2 files will be created**: a zip file and pdf file.  You can download them using the links provided below OR by finding them in the same folder where this juptyer notebook resides in your JuptyerHub.

To receive credit on this assignment, **you must submit BOTH of these files
to their respective Gradescope portals:** 

* **Homework 3 Autograded**: Submit the zip file that is output by the `grader.export()` cell below to the HW1 Autograded assignment in Gradescope.

* **Homework 3 Manually Graded**: Submit your hw01.PDF to the HW1 Manually Graded assignment in Gradescope.  


**You are responsible for ensuring your submission follows our requirements. We will not be granting regrade requests nor extensions to submissions that don't follow instructions.** If you encounter any difficulties with submission, please don't hesitate to reach out to the instructor prior to the deadline.

## Submission

Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. The cell below will generate a zip file for you to submit. **Please save before exporting!**

AFTER running the cell below, click on <a href='hw03.pdf' download>this link to download the PDF </a> to upload to Gradescope.  There will be a separate link that appears after running the cell below with a link to download the zip file to upload to Gradescope.

In [ ]:
# Save your notebook first, then run this cell to export your submission.
grader.export(run_tests=True)